File: python/composio/core/models/tool_router_session_api.py
"""Small v3.1 Tool Router helpers used until generated clients expose the routes."""

from __future__ import annotations

import typing as t
from urllib.parse import quote

from composio_client.types.tool_router.session_tools_response import (
    SessionToolsResponse,
)

from composio.client import HttpClient

SESSION_TOOLS_PAGE_LIMIT = 500


def list_all_tool_router_session_tools_v31(
    client: HttpClient,
    session_id: str,
) -> t.List[t.Any]:
    tools: t.List[t.Any] = []
    cursor: t.Optional[str] = None
    encoded_session_id = quote(session_id, safe="")

    while True:
        params: t.Dict[str, t.Any] = {"limit": SESSION_TOOLS_PAGE_LIMIT}
        if cursor:
            params["cursor"] = cursor

        response = t.cast(
            SessionToolsResponse,
            client.get(
                f"/api/v3.1/tool_router/session/{encoded_session_id}/tools",
                cast_to=SessionToolsResponse,
                options={"params": params},
            ),
        )
        tools.extend(response.items)
        cursor = getattr(response, "next_cursor", None)
        if not cursor:
            return tools

File: ts/packages/core/src/lib/toolRouterSessionApi.ts
import type { Composio as ComposioClient } from '@composio/client';
import type {
  SessionCreateParams,
  SessionCreateResponse,
  SessionToolsResponse,
} from '@composio/client/resources/tool-router/session/session.mjs';

const TOOL_ROUTER_SESSION_TOOLS_PAGE_LIMIT = 500;
const encodePathSegment = (value: string): string => encodeURIComponent(value);

export type SessionCreateParamsV31 = SessionCreateParams & {
  search?: {
    enable?: boolean;
  };
  execution?: {
    enable_multi_execute?: boolean;
  };
};

export type SessionToolsResponseV31 = SessionToolsResponse & {
  next_cursor?: string | null;
  total_pages?: number;
};

export const createToolRouterSessionV31 = (
  client: ComposioClient,
  payload: SessionCreateParamsV31
): Promise<SessionCreateResponse> => {
  return client.post('/api/v3.1/tool_router/session', { body: payload });
};

const listToolRouterSessionToolsV31 = (
  client: ComposioClient,
  sessionId: string,
  query?: { cursor?: string | null; limit?: number }
): Promise<SessionToolsResponseV31> => {
  return client.get(`/api/v3.1/tool_router/session/${encodePathSegment(sessionId)}/tools`, {
    query,
  });
};

export const listAllToolRouterSessionToolsV31 = async (
  client: ComposioClient,
  sessionId: string
): Promise<SessionToolsResponse['items']> => {
  const items: SessionToolsResponse['items'] = [];
  let cursor: string | null | undefined;

  do {
    const response = await listToolRouterSessionToolsV31(client, sessionId, {
      limit: TOOL_ROUTER_SESSION_TOOLS_PAGE_LIMIT,
      ...(cursor ? { cursor } : {}),
    });
    items.push(...response.items);
    cursor = response.next_cursor;
  } while (cursor);

  return items;
};
